### PR TITLE
Fixing wmdistance

### PR DESCRIPTION
**Pull request by ibrahimsharaf**
Issue (https://github.com/marco-c/crashsimilarity/issues/61)
- Refactoring wmdistance, solving ValueError exception.
- Now, you can calculate wmdistance using euclidean distance or cosine similarity.
- Adding 2 wmdistance tests (zero, non zero distance).

### CODE REVIEW

**Comment by ibrahimsharaf** [comment]
I think the wmd refactoring has introduced a problem in test_zero_dist_coherence which uses signature_similarity in crash_similarity.py, should I report a new issue? @marco-c

File: crashsimilarity/crash_similarity.py [hunk 1]
```diff
@@ -10,14 +10,19 @@
...
-from pyemd import emd
 
 from crashsimilarity import utils
 from crashsimilarity.downloader import SocorroDownloader
 from crashsimilarity.utils import StackTraceProcessor
 
 pyximport.install()
 
+try:
+    from pyemd import emd
```
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: Let's make this package strictly required as it was before, it's in the requirements.txt file anyway

File: tests/test_crash_similarity.py [hunk 2]
```diff
@@ -85,6 +86,39 @@ def test_train_model(self):
         self.assertEqual(20, resp.iter)
         self.assertEqual(101, len(resp.wv.vocab))
 
+    def test_wmdistance_non_zero_distance(self):
```
Comment (marco-c) [inline tests/test_crash_similarity.py]: Nit: move this right after `test_order_similarity`, this way we keep all model-related tests together.

**Comment by marco-c** [comment]
> I think the wmd refactoring has introduced a problem in test_zero_dist_coherence which uses signature_similarity in crash_similarity.py, should I report a new issue? @marco-c

If the test is now passing, you can simply remove the `@unittest.expectedFailure` annotation from the test.

File: crashsimilarity/crash_similarity.py [hunk 3]
```diff
@@ -89,40 +94,63 @@ def train_model(corpus, force_train=False):
...
-
-
 # Code modified from https://github.com/RaRe-Technologies/gensim/blob/4f0e2ae/gensim/models/keyedvectors.py#L339
-def wmdistance(model, words1, words2, all_distances):
-    dictionary = gensim.corpora.Dictionary(documents=[words1, words2])
+def wmdistance(model, document1, document2, all_distances, distance_metric='cosine'):
+    if not PYEMD_EXT:
+        raise ImportError("Please install pyemd Python package to compute WMD.")
+
+    # Remove out-of-vocabulary words.
```
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: The out-of-vocabulary words are already removed in the caller

File: crashsimilarity/crash_similarity.py [hunk 4]
```diff
@@ -89,40 +94,63 @@ def train_model(corpus, force_train=False):
...
-    bow1 = create_bow(words1)
-    bow2 = create_bow(words2)
+    if np.sum(distance_matrix) == 0.0:
+        # `emd` gets stuck if the distance matrix contains only zeros.
+        logging.info('The distance matrix is all zeros. Aborting (returning inf).')
+        return float('inf')
 
-    docset = set(words2)
-    distances = create_distance_matrix(model, dictionary, docset, all_distances)
+    def nbow(document):
```
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: Did you do this make it more similar to the GenSim code? It's worth doing it, but it should be a separate PR.
Each PR should be self-contained.

**Comment by marco-c** [comment]
Let's split the PR in three PRs: one to fix the zero-distance issue and adding the tests; one to rename the variables to make them more similar to the GenSim code; one to add the euclidean distance option.

File: crashsimilarity/crash_similarity.py (hunk 4 above)
Comment (ibrahimsharaf) [inline crashsimilarity/crash_similarity.py]: Yes, will do it.
